--- /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/config/application.py	2013-04-06 05:21:31.000000000 +1000
+++ Contents/Resources/opt/local/bin/Python.framework.2.7/lib/python2.7/site-packages/IPython/config/application.py	2014-01-09 12:54:08.000000000 +1000
@@ -64,7 +64,7 @@
 subcommand_description = """
 Subcommands are launched as `{app} cmd [args]`. For information on using
 subcommand 'cmd', do: `{app} cmd -h`.
-""".strip().format(app=os.path.basename(sys.argv[0]))
+""".strip().format(app=os.path.basename("/tmp"))
 # get running program name
 
 #-----------------------------------------------------------------------------
